cpp/tests/iterator/iterator_test.cu

```diff
@@ -0,0 +1,393 @@
+#include <thrust/transform.h>
+
+#include <thrust/tuple.h>
+#include <thrust/iterator/zip_iterator.h>
+#include <thrust/iterator/constant_iterator.h>
+#include <thrust/iterator/counting_iterator.h>
+#include <thrust/iterator/transform_iterator.h>
+#include <thrust/iterator/iterator_adaptor.h>
+
+
+// --------------------------------------------------------------------------------------------------------
+// structs for column_input_iterator
+template<typename T>
+struct ColumnDataNonNull{
+    const T *data;
+
+    __device__ __host__
+    ColumnDataNonNull(T *_data)
+    : data(_data){};
+
+    __device__ __host__
+    T at(gdf_index_type id) const {
+        return data[id];
+    };
+};
+
+template<typename T>
+struct ColumnData{
+    const T *data;
+    const gdf_valid_type *valid;
+    T identity;
+
+    __device__ __host__
+    ColumnData(T *_data, gdf_valid_type *_valid, T _identity)
+    : data(_data), valid(_valid), identity(_identity){};
+
+    __device__ __host__
+    T at(gdf_index_type id) const {
+        return (gdf_is_valid(valid, id))? data[id] : identity;
+    };
+};
+
+template<typename T>
+struct ColumnDataSquare : public ColumnData<T>
+{
+    ColumnDataSquare(T *_data, gdf_valid_type *_valid, T _identity)
+    : ColumnData<T>(_data, _valid, _identity){};
+
+    __device__ __host__
+    T at(gdf_index_type id) const {
+        T val = ColumnData<T>::at(id);
+        return (val * val);
+    };
+};
+
+// TBD. ColumnDataSquareNonNull
+
+
+// column_input_iterator
+template<typename T_output, typename T_input, typename Iterator=thrust::counting_iterator<gdf_index_type> >
+  class column_input_iterator
+    : public thrust::iterator_adaptor<
+        column_input_iterator<T_output, T_input, Iterator>, // the first template parameter is the name of the iterator we're creating
+        Iterator,                   // the second template parameter is the name of the iterator we're adapting
+        thrust::use_default, thrust::use_default, thrust::use_default, T_output, thrust::use_default
+      >
+      {
+  public:
+    // shorthand for the name of the iterator_adaptor we're deriving from
+    using super_t = thrust::iterator_adaptor<
+      column_input_iterator<T_output, T_input, Iterator>,
+      Iterator,
+      thrust::use_default, thrust::use_default, thrust::use_default, T_output, thrust::use_default
+    >;
+
+    __host__ __device__
+    column_input_iterator(const T_input col, const Iterator &it) : super_t(it), colData(col){}
+
+    __host__ __device__
+    column_input_iterator(const T_input col) : super_t(Iterator{0}), colData(col){}
+
+    __host__ __device__
+    column_input_iterator(const column_input_iterator &it) : super_t(it.base()), colData(it.colData){}
+
+    // befriend thrust::iterator_core_access to allow it access to the private interface below
+    friend class thrust::iterator_core_access;
+
+    __host__ __device__
+    column_input_iterator &operator=(const column_input_iterator &other)
```

Comment:
Is this necessary?

Reply:
I',m still investigating if it is necessary.
Actually, I've referenced the implementation of `thrust::transform_iterator`.
https://github.com/thrust/thrust/blob/cc4be34c4b3bf7a4b5a7ef2977bab5bfcc676878/thrust/iterator/transform_iterator.h#L245-L297